src/solidity/stateDecoder.js

```diff
@@ -0,0 +1,233 @@
+'use strict'
+var BN = require('ethereumjs-util').BN
+var utileth = require('ethereumjs-util')
+var varUtil = require('./variable')
+
+module.exports = {
+  decodeArray: function (type, storage, location) {
+    return this.decodeArrayItems(type, storage, 0, location)
+  },
+
+  decodeArrayItems: function (type, storage, depth, location) {
+    var ret = []
+    var size = type.dim[depth]
+    if (size === 'dynamic') {
+      size = storage[formatHexKey(location.slot.toString(16))]
+      size = parseInt(size)
+      var pointer = getDynamicPointer(location)
+      location.offset = 0
+      location.slot = new BN(pointer.replace('0x', ''), 16)
+    }
+    if (type.dim.length - 1 > depth) {
+      depth++
+      for (var k = 0; k < size; k++) {
+        ret.push(this.decodeArrayItems(type, storage, depth, location))
+        location = this.moveNextArray(location)
+      }
+    } else {
+      for (var i = 0; i < size; i++) {
+        ret.push(this.decodeType(type, storage, location))
+        location = this.moveNextItemInArray(type, location)
+      }
+    }
+    return ret
+  },
+
+  moveNextArray: function (location) {
+    return {
+      offset: 0,
+      slot: location.slot + 1
+    }
+  },
+
+  moveNextItemInArray: function (type, location) {
+    if (type.size === 'dynamic' || location.offset + type.memSize >= 64 || type.isStruct) {
+      location.offset = 0
+      location.slot = add(location.slot, 1)
+    } else {
+      location.offset = location.offset + type.memSize
+    }
+    return location
+  },
+
+  decodeInt: function (type, storage, location) {
+    var value = getValue(type, storage, location)
+    value = extractValue(value, type, location)
+    var bigNumber = new BN(value.replace('0x', ''), 16)
+    if (type.innerType === 'uint') {
+      return bigNumber.toString(10)
+    } else if (type.innerType === 'int') {
+      if (this.isNegative(value)) {
+        return '-' + bigNumber.sub(new BN(new Array(value.length + 1).join('f'), 16)).sub(new BN(1)).toString(10)
```

Comment:
yes there might be a better way to do that... but i can't figure out it using BigNumber 
we need to subtract using a ''fffff' that has the same length as the value,
and the length of value can be everything from 2 to 64. 
another example is there:
https://github.com/ethereum/web3.js/blob/develop/lib/utils/utils.js#L377

https://github.com/ethereum/web3.js/blob/develop/lib/utils/utils.js#L377